server/server.go

```diff
@@ -1229,11 +1249,112 @@ func (server *ArgoCDServer) newHTTPServer(ctx context.Context, port int, grpcWeb
 	// Dex reverse proxy and client app and OAuth2 login/callback
 	server.registerDexHandlers(mux)
 
+	namespace, _, err := server.ArgoCDServerOpts.ClientConfig.Namespace()
+	if err != nil {
+		log.Error(err, "error while getting namespace from client config")
+		os.Exit(1)
+	}
+	server.ArgoCDServerOpts.ApplicationNamespaces = append(server.ArgoCDServerOpts.ApplicationNamespaces, namespace)
+
 	// Webhook handler for git events (Note: cache timeouts are hardcoded because API server does not write to cache and not really using them)
 	argoDB := db.NewDB(server.Namespace, server.settingsMgr, server.KubeClientset)
-	acdWebhookHandler := webhook.NewHandler(server.Namespace, server.ApplicationNamespaces, server.WebhookParallelism, server.AppClientset, server.settings, server.settingsMgr, server.RepoServerCache, server.Cache, argoDB, server.settingsMgr.GetMaxWebhookPayloadSize())
+	acdWebhookHandler := argocdwebhook.NewHandler(server.Namespace, server.ApplicationNamespaces, server.WebhookParallelism, server.AppClientset, server.settings, server.settingsMgr, server.RepoServerCache, server.Cache, argoDB, server.settingsMgr.GetMaxWebhookPayloadSize())
+
+	cfg := ctrl.GetConfigOrDie()
+	err = appv1alpha1.SetK8SConfigDefaults(cfg)
+	if err != nil {
+		log.Error(err, "Unable to apply K8s REST config defaults")
+		os.Exit(1)
+	}
+
+	scheme := runtimepkg.NewScheme()
+	_ = clientgoscheme.AddToScheme(scheme)
+	_ = appv1alpha1.AddToScheme(scheme)
+
+	// By default, watch all namespaces
+	var watchedNamespace string
+	// If the applicationset-namespaces contains only one namespace it corresponds to the current namespace
+	if len(server.ApplicationSetOpts.ApplicationSetNamespaces) == 1 {
+		watchedNamespace = (server.ApplicationSetOpts.ApplicationSetNamespaces)[0]
+	} else if server.ApplicationSetOpts.EnableScmProviders && len(server.ApplicationSetOpts.AllowedScmProviders) == 0 {
+		log.Error("When enabling applicationset in any namespace using applicationset-namespaces, you must either set --enable-scm-providers=false or specify --allowed-scm-providers")
+		os.Exit(1)
+	}
+
+	var cacheOpt ctrlcache.Options
+	if watchedNamespace != "" {
+		cacheOpt = ctrlcache.Options{
+			DefaultNamespaces: map[string]ctrlcache.Config{
+				watchedNamespace: {},
+			},
+		}
+	}
+
+	mgr, err := ctrl.NewManager(cfg, ctrl.Options{
+		Scheme: scheme,
+		Metrics: metricsserver.Options{
+			BindAddress: server.ApplicationSetOpts.MetricsAddr,
+		},
+		Cache:                  cacheOpt,
+		HealthProbeBindAddress: server.ApplicationSetOpts.ProbeBindAddr,
+		LeaderElection:         server.ApplicationSetOpts.EnableLeaderElection,
+		LeaderElectionID:       "58ac56fa.applicationsets.argoproj.io",
+		Client: ctrlclient.Options{
+			DryRun: &server.ApplicationSetOpts.DryRun,
+		},
+	})
+	if err != nil {
+		log.Error(err, "unable to start manager")
+		os.Exit(1)
+	}
+	dynamicClient, err := dynamic.NewForConfig(mgr.GetConfig())
```

Comment:
I wonder if the two-step of building a manager and then constructing a client from the manager's config can be avoided. That would allow us to eliminate a ton of config dependencies.